tests/kafkatest/services/kafka/kafka.py

```diff
@@ -165,6 +165,19 @@ def start(self, add_principals=""):
         self.start_minikdc(add_principals)
         Service.start(self)
 
+        self.logger.info("Waiting for brokers to register at ZK")
+
+        cmd = self.path.script("zookeeper-shell.sh", self.zk) + " " + self.zk_connect_setting() + " ls /brokers/ids"
+
+        broker_ids = ""
+        for node in self.nodes:
+            broker_ids += "%s, " % self.idx(node)
+        broker_ids = "[" + broker_ids[:-2] + "]"
+
+        with self.zk.account.monitor_log(KafkaService.STDOUT_STDERR_CAPTURE) as monitor:
+            self.zk.account.ssh(cmd)
+            monitor.wait_until(broker_ids, timeout_sec=30, backoff_sec=.25, err_msg="Kafka servers didn't register at ZK")
```

Comment:
i don't think this is what you want here. `broker_ids` is evaluated once. the `wait_until` implies you want to wait until the `broker_ids` appear in the the log files for the broker. but that doesn't seem like the target you want.

Reply:
I am still learning about ducktape... How can I trigger the ZK command multiple times and check if it outputs what I want to see up to a max retry/timeout? It there any helper method I can use? Or do I need to do a manual loop?

Reply:
i don't think there's a good way to do this directly with `monitor.wait_until` because it was written originally to look for a fixed pattern (we've mostly used it to monitor startup where we know a specific message gets logged once the service is actually serving requests). you could either look for the fixed pattern (e.g. if this message only gets logged once and before the broker ids it has a known msg) or do a `wait_until` yourself. `wait_until` takes a function, so you can just wrap up the check in a local function that returns a boolean and pass that into `wait_until`.